File: src/main/java/com/jewelexx/tablocation/TabLocation.java
package com.jewelexx.tablocation;

import java.util.logging.Logger;

import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.World.Environment;
import org.bukkit.block.Block;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerMoveEvent;
import org.bukkit.plugin.PluginManager;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.scoreboard.Scoreboard;
import org.bukkit.scoreboard.Team;

import com.jewelexx.craftcolours.CraftColours;

enum ShowDimension {
    Minimal,
    Expanded,
    False,
}

public final class TabLocation extends JavaPlugin implements Listener {
    static final Logger log = Bukkit.getLogger();
    static ShowDimension environmentEnabled;
    static boolean locationEnabled;
    static boolean bracketColourEnabled;
    static FileConfiguration config;
    final String version = getDescription().getVersion();

    @Override
    public void onEnable() {
        new org.bstats.bukkit.Metrics(this, 9922);

        PluginManager manager = Bukkit.getPluginManager();

        saveDefaultConfig();

        config = getConfig();

        String showDimension = config.getString("Show dimension");

        if (showDimension == "minimal") {
            environmentEnabled = ShowDimension.Minimal;
        } else if (showDimension == "true") {
            environmentEnabled = ShowDimension.Expanded;
        } else {
            environmentEnabled = ShowDimension.False;
        }

        locationEnabled = config.getBoolean("Show location");
        bracketColourEnabled = config.getBoolean("Colour brackets");

        manager.registerEvents(this, this);

        if (manager.getPlugin("PlaceholderAPI") != null) {
            new Placeholders(this).register();
        }

        log.info("===================================");
        log.info("TabLocation has been enabled!");
        log.info("Version " + version);
        log.info("Developed with 💗 by Juliette Cordor");
        log.info("===================================");

        new UpdateChecker(this).getVersion(version -> {
            if (!getDescription().getVersion().equals(version)) {
                log.warning("[TabLocation] There is a new update available.");
            }
        });
    }

    @Override
    public void onDisable() {
        log.info("===================================");
        log.info("Plugin has been disabled!");
        log.info("Thank you for using TabLocation!");
        log.info("===================================");
    }

    @EventHandler
    static void onPlayerMove(PlayerMoveEvent e) {
        Block from = e.getFrom().getBlock();
        Block to = e.getTo().getBlock();
        if (from.getX() != to.getX() || from.getY() != to.getY() || from.getZ() != to.getZ()) {
            updateLocation(e.getPlayer());
        }
    }

    @EventHandler
    static void onPlayerJoin(PlayerJoinEvent e) {
        updateLocation(e.getPlayer());
    }

    static String withTeamName(Player player) {
        String name = player.getDisplayName();
        Scoreboard sb = Bukkit.getScoreboardManager().getMainScoreboard();

        for (Team team : sb.getTeams()) {
            if (team.hasEntry(player.getName())) {
                return team.getColor() + team.getPrefix() + name + team.getSuffix() + CraftColours.RESET;
            }
        }

        return name;
    }

    static void updateLocation(Player player) {
        player.setPlayerListName(withTeamName(player) + getLoc(player));
    }

    protected static String getLoc(Player player) {
        if ((!locationEnabled && environmentEnabled == ShowDimension.False)
                || player.hasPermission("tablocation.hide")) {
            return "";
        }

        String colourcode = CraftColours.WHITE;

        String world = "";

        if (environmentEnabled != ShowDimension.False) {
            Environment environment = player.getWorld().getEnvironment();

            switch (environment) {
                case NORMAL:
                    world = "Overworld";
                    break;
                case NETHER:
                    world = "Nether";
                    break;
                case THE_END:
                    world = "End";
                    break;
                default:
                    world = environment.toString();
                    break;
            }

            colourcode = config.getString("Colour for The " + world);

            if (environmentEnabled == ShowDimension.Expanded) {
                world = colourcode + "The " + world + CraftColours.RESET;
            }
        }

        String location = "";

        if (locationEnabled) {
            Location loc = player.getLocation();
            int x = loc.getBlockX();
            int y = loc.getBlockY();
            int z = loc.getBlockZ();
            location = x + ", " + y + ", " + z;
        }

        String separator = "";

        if (locationEnabled && environmentEnabled == ShowDimension.Expanded) {
            separator = ", ";
        }

        return " " + colourcode + "[" + CraftColours.WHITE + location + separator + world + colourcode + "]";
    }
}

File: src/main/kotlin/dev/cordor/tablocation/Tablocation.kt
package dev.cordor.tablocation

import com.jewelexx.craftcolours.CraftColours
import io.papermc.paper.ServerBuildInfo
import net.kyori.adventure.audience.Audience
import net.kyori.adventure.key.Key
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.ComponentBuilder
import net.kyori.adventure.text.TextComponent
import org.bukkit.Bukkit
import org.bukkit.World.Environment
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerMoveEvent
import org.bukkit.plugin.java.JavaPlugin
import java.net.URI


enum class ShowDimension {
    Minimal,
    Expanded,
    False,
}

class Tablocation : JavaPlugin(), Listener {
    lateinit var environmentEnabled: ShowDimension
    var locationEnabled: Boolean = false
    val version = description.version

    companion object {
        fun isFolia(): Boolean {
            return ServerBuildInfo.buildInfo().isBrandCompatible(Key.key("papermc", "folia"))
        }
    }

    override fun onEnable() {
        org.bstats.bukkit.Metrics(this, 9922)

        saveDefaultConfig()
        val showDimension = config.getString("Show dimension")

        environmentEnabled = when (showDimension) {
            "minimal" -> {
                ShowDimension.Minimal
            }

            "expanded" -> {
                ShowDimension.Expanded
            }

            else -> {
                ShowDimension.False
            }
        }
        locationEnabled = config.getBoolean("Show location")

        val manager = server.pluginManager

        manager.registerEvents(this, this)
        if (manager.getPlugin("PlaceholderAPI") != null) {
            Placeholders(this).register()
        }

        logger.info("===================================")
        logger.info("TabLocation has been enabled!")
        logger.info("Version $version")
        logger.info("Developed with 💗 by Juliette Cordor")
        logger.info("===================================")

        UpdateChecker(this, URI("https://api.github.com/repos/jewlexx/TabLocation/tags").toURL()).getVersion { version ->
            if (description.version != version) {
                logger.warning("[TabLocation] There is a new update available!")
            }
        }
    }

    override fun onDisable() {
        logger.info("===================================")
        logger.info("Plugin has been disabled!")
        logger.info("Thank you for using TabLocation!")
        logger.info("===================================")
    }

    @EventHandler
    fun onPlayerMove(e: PlayerMoveEvent) {
        val fromBlock = e.from.block
        val toBlock = e.to.block ?: return

        if (fromBlock.x != toBlock.x || fromBlock.y != toBlock.y || fromBlock.z != toBlock.z) {
            updateLocation(e.player)
        }
    }

    @EventHandler
    fun onPlayerJoin(e: PlayerJoinEvent) {
        updateLocation(e.player)
    }

    fun updateLocation(player: Player) {
        val newName = player.playerListName();
        newName.append(Component.text(getLoc(player)));
        player.playerListName(newName)
    }

//    fun withTeamName(player: Player): String {
//        val name = player.displayName()
//        val sb = server.scoreboardManager.mainScoreboard
//
//        for (team in sb.teams) {
//            if (team.hasEntry(player.name)) {
//                val nameComponent = Component.empty()
//                nameComponent.append(team.name)
//                return "${team.color}${team.prefix}$name${team.suffix}${CraftColours.RESET}"
//            }
//        }
//
//        return name
//    }

    fun getLoc(player: Player): String {
        if ((!locationEnabled && environmentEnabled == ShowDimension.False)
            || player.hasPermission("tablocation.hide")
        ) {
            return ""
        }

        var colourCode = CraftColours.WHITE

        val world = if (environmentEnabled != ShowDimension.False) {
            val environmentName = when (val environment = player.world.environment) {
                Environment.NORMAL -> "Overworld"
                Environment.NETHER -> "Nether"
                Environment.THE_END -> "End"
                else -> environment.toString()
            }

            val dimensionColourCode = config.getString("Colour for The $environmentName") ?: ""

            if (environmentEnabled == ShowDimension.Expanded) {
                "${dimensionColourCode}The $environmentName${CraftColours.RESET}"
            } else {
                colourCode = dimensionColourCode
                // Hide `world` variable if displaying minimal
                ""
            }
        } else {
            ""
        }

        val location = if (locationEnabled) {
            val loc = player.location
            val x = loc.blockX
            val y = loc.blockY
            val z = loc.blockZ
            "$x, $y, $z"
        } else {
            ""
        }

        val separator = if (locationEnabled && environmentEnabled == ShowDimension.Expanded) {
            ", "
        } else {
            ""
        }

        return " $colourCode[${CraftColours.WHITE}$location$separator$world$colourCode]"
    }
}
